File: app/src/main/java/uk/co/ribot/androidboilerplate/data/model/bean/Subject.java
package uk.co.ribot.androidboilerplate.data.model.bean;

import android.content.ContentValues;
import android.database.Cursor;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import com.google.auto.value.AutoValue;
import com.google.gson.Gson;
import com.google.gson.TypeAdapter;
import com.google.gson.reflect.TypeToken;
import com.squareup.sqldelight.ColumnAdapter;
import com.squareup.sqldelight.RowMapper;
import com.sunny.sql.SubjectModel;

import java.util.List;

import uk.co.ribot.androidboilerplate.data.model.pojo.Image;
import uk.co.ribot.androidboilerplate.data.model.pojo.Person;
import uk.co.ribot.androidboilerplate.data.model.pojo.Rating;

/**
 * Created by Zhou Zejin on 2016/9/21.
 */

@AutoValue
public abstract class Subject implements SubjectModel, Parcelable, Comparable<Subject> {

    private static final ColumnAdapter RATING_ADAPTER = new ColumnAdapter<Rating>() {
        @Override
        public Rating map(Cursor cursor, int columnIndex) {
            return new Gson().fromJson(cursor.getColumnName(columnIndex), Rating.class);
        }

        @Override
        public void marshal(ContentValues values, String key, Rating value) {
            values.put(key, new Gson().toJson(value));
        }
    };

    private static final ColumnAdapter GENRES_ADAPTER = new ColumnAdapter<List<String>>() {
        @Override
        public List<String> map(Cursor cursor, int columnIndex) {
            return new Gson().fromJson(
                    cursor.getColumnName(columnIndex),
                    new TypeToken<List<String>>() {
                    }.getType());
        }

        @Override
        public void marshal(ContentValues values, String key, List<String> value) {
            values.put(key, new Gson().toJson(value));
        }
    };

    private static final ColumnAdapter PERSONS_ADAPTER = new ColumnAdapter<List<Person>>() {
        @Override
        public List<Person> map(Cursor cursor, int columnIndex) {
            return new Gson().fromJson(
                    cursor.getColumnName(columnIndex),
                    new TypeToken<List<String>>() {
                    }.getType());
        }

        @Override
        public void marshal(ContentValues values, String key, List<Person> value) {
            values.put(key, new Gson().toJson(value));
        }
    };

    private static final ColumnAdapter IMAGE_ADAPTER = new ColumnAdapter<Image>() {
        @Override
        public Image map(Cursor cursor, int columnIndex) {
            return new Gson().fromJson(cursor.getColumnName(columnIndex), Image.class);
        }

        @Override
        public void marshal(ContentValues values, String key, Image value) {
            values.put(key, new Gson().toJson(value));
        }
    };

    public static final Factory<Subject> FACTORY = new Factory<>(
            new SubjectModel.Creator<Subject>() {
                @Override
                public Subject create(@NonNull String id, @NonNull Rating rating,
                                      @NonNull List<String> genres, @NonNull String title,
                                      @NonNull List<Person> casts, int collect_count,
                                      @NonNull String original_title, @NonNull String subtype,
                                      @NonNull List<Person> directors, @NonNull String year,
                                      @NonNull Image images, @NonNull String alt) {
                    return new AutoValue_Subject(id, rating, genres, title, casts, collect_count,
                            original_title, subtype, directors, year, images, alt);
                }
            },
            RATING_ADAPTER,
            GENRES_ADAPTER,
            PERSONS_ADAPTER,
            PERSONS_ADAPTER,
            IMAGE_ADAPTER);

    public static final RowMapper<Subject> MAPPER = FACTORY.select_allMapper();

    public static Builder builder() {
        return new AutoValue_Subject.Builder();
    }

    @AutoValue.Builder
    public abstract static class Builder {
        public abstract Builder id(String id);

        public abstract Builder rating(Rating rating);

        public abstract Builder genres(List<String> genres);

        public abstract Builder title(String title);

        public abstract Builder casts(List<Person> casts);

        public abstract Builder collect_count(int collect_count);

        public abstract Builder original_title(String original_title);

        public abstract Builder subtype(String subtype);

        public abstract Builder directors(List<Person> directors);

        public abstract Builder year(String year);

        public abstract Builder images(Image images);

        public abstract Builder alt(String alt);

        public abstract Subject build();
    }

    public static TypeAdapter<Subject> typeAdapter(Gson gson) {
        return new AutoValue_Subject.GsonTypeAdapter(gson);
    }

    @Override
    public int compareTo(@NonNull Subject another) {
        return title().compareTo(another.title());
    }

}

File: app/src/main/java/uk/co/ribot/androidboilerplate/data/model/bean/Subject.kt
package uk.co.ribot.androidboilerplate.data.model.bean

import android.content.ContentValues
import android.database.Cursor
import android.os.Parcelable

import com.gabrielittner.auto.value.cursor.ColumnAdapter
import com.google.auto.value.AutoValue
import com.google.gson.Gson
import com.google.gson.TypeAdapter
import com.squareup.sqldelight.RowMapper
import com.sunny.sql.SubjectModel

import uk.co.ribot.androidboilerplate.data.model.pojo.Image
import uk.co.ribot.androidboilerplate.data.model.pojo.Person
import uk.co.ribot.androidboilerplate.data.model.pojo.Rating
import uk.co.ribot.androidboilerplate.utils.factory.MyColumnAdapterFactory
import uk.co.ribot.androidboilerplate.utils.factory.MyColumnTypeAdapterFactory

/**
 * Created by Zhou Zejin on 2016/9/21.
 */

@AutoValue
abstract class Subject : SubjectModel, Parcelable, Comparable<Subject> {

    abstract override fun id(): String

    @ColumnAdapter(MyColumnTypeAdapterFactory.RatingAdapter::class)
    abstract override fun rating(): Rating

    @ColumnAdapter(MyColumnTypeAdapterFactory.ListStringAdapter::class)
    abstract override fun genres(): List<String>

    abstract override fun title(): String

    @ColumnAdapter(MyColumnTypeAdapterFactory.ListPersonAdapter::class)
    abstract override fun casts(): List<Person>

    abstract override fun collect_count(): Int

    abstract override fun original_title(): String

    abstract override fun subtype(): String

    @ColumnAdapter(MyColumnTypeAdapterFactory.ListPersonAdapter::class)
    abstract override fun directors(): List<Person>

    abstract override fun year(): String

    @ColumnAdapter(MyColumnTypeAdapterFactory.ImageAdapter::class)
    abstract override fun images(): Image

    abstract override fun alt(): String

    @AutoValue.Builder
    abstract class Builder {
        abstract fun id(id: String): Builder
        abstract fun rating(rating: Rating): Builder
        abstract fun genres(genres: List<String>): Builder
        abstract fun title(title: String): Builder
        abstract fun casts(casts: MutableList<Person>): Builder
        abstract fun collect_count(collect_count: Int): Builder
        abstract fun original_title(original_title: String): Builder
        abstract fun subtype(subtype: String): Builder
        abstract fun directors(directors: MutableList<Person>): Builder
        abstract fun year(year: String): Builder
        abstract fun images(images: Image): Builder
        abstract fun alt(alt: String): Builder
        abstract fun build(): Subject
    }

    abstract fun toContentValues(): ContentValues

    override fun compareTo(other: Subject): Int {
        return title().compareTo(other.title())
    }

    companion object {
        val FACTORY: SubjectModel.Factory<Subject> = SubjectModel.Factory(
                SubjectModel.Creator<Subject> { id, rating, genres, title, casts, collect_count,
                                                original_title, subtype, directors, year, images, alt ->
                    AutoValue_Subject(id, rating, genres, title, casts, collect_count,
                            original_title, subtype, directors, year, images, alt)
                },
                MyColumnAdapterFactory.RATING_ADAPTER,
                MyColumnAdapterFactory.GENRES_ADAPTER,
                MyColumnAdapterFactory.PERSONS_ADAPTER,
                MyColumnAdapterFactory.PERSONS_ADAPTER,
                MyColumnAdapterFactory.IMAGE_ADAPTER)

        val MAPPER: RowMapper<Subject> = FACTORY.selectAllMapper()

        @JvmStatic
        fun create(cursor: Cursor): Subject {
            return `$$AutoValue_Subject`.createFromCursor(cursor)
        }

        @JvmStatic
        fun typeAdapter(gson: Gson): TypeAdapter<Subject> {
            return `$$$AutoValue_Subject`.GsonTypeAdapter(gson)
        }

        fun builder(): Builder {
            return `$$$$AutoValue_Subject`.Builder()
        }
    }

}
